mobilecoind-dev-faucet/src/worker.rs

```diff
@@ -570,83 +590,406 @@ impl WorkerTokenState {
+            //   also be less than the average. So the TXOs produced by this rebalancing
+            //   will be the top utxos after this rebalancing Tx lands, and they are all
+            //   nearly equal.
+            // * The avg_value also will not change much.
             //
-            // FIXME: We should also detect the situation that defragmentation is required
-            // and then submit defragmentation txs.
-            if non_target_value_utxos
-                .iter()
-                .map(|utxo| utxo.value)
-                .sum::<u64>()
-                < self.target_value * (MAX_OUTPUTS - 1) + self.minimum_fee_value
-            {
-                let prev_value = self.funds_depleted.swap(true, Ordering::SeqCst);
-                if !prev_value {
-                    log::info!(logger, "Funds depleted on {}", self.token_id);
-                }
-                return Ok(());
-            } else {
-                let prev_value = self.funds_depleted.swap(false, Ordering::SeqCst);
-                if prev_value {
-                    log::info!(logger, "Funds no longer depleted on {}", self.token_id);
-                }
+            // So we will likely not meet the criteria after this rebalancing
+            // operation, and there will not be an infinite loop of rebalancing
+            // operations which don't refill the queue.
+            log::info!(
+                logger,
+                "Attempting a rebalancing Tx for split tx parallelism on token id {}",
+                self.token_id
+            );
+
+            // Check if any of these UTXOs were used by an in-flight split tx or
+            // defrag tx. If so then we
+            // should back off and wait for it to clear and re-evaluate.
+            if top_utxos.iter().any(|utxo| {
+                let key_image = utxo.get_key_image().try_into().unwrap();
+                self.in_flight_split_tx_states.contains_key(&key_image)
+                    || self
+                        .in_flight_defragmentation_key_images
+                        .contains(&key_image)
+            }) {
+                log::info!(
+                    logger,
+                    "Backing off before sending a rebalancing tx {}",
+                    self.token_id
+                );
+                return Ok(false);
             }
 
             // Generate an outlay
             // We will repeat this outlay MAX_OUTPUTS - 1 times
-            // (-1 is for a change output)
+            // (-1 is for a change output, which might be slightly larger than avg_value, or
+            // less due to fees)
             let mut outlay = api::Outlay::new();
             outlay.set_receiver(public_address.clone());
-            outlay.set_value(self.target_value);
+            outlay.set_value(avg_value);
 
             // Generate a Tx
+            // Note: This will fail if MAX_INPUTS < MAX_OUTPUTS, but right now MAX_INPUTS =
+            // MAX_OUTPUTS.
```

Comment:
Might it make sense `assert` this precondition? Or wouldthe failure be self-explanatory?

Reply:
we could assert this, I'm not sure how clear the reason would be if we don't. It might be something like "input selection failed"